### PR TITLE
Split pytest plugins out into individual modules

### PR DESCRIPTION
**Pull request by Cadair**
This makes it possible for affiliated packages (SunPy) to use some of these options but not all of them.

### CODE REVIEW

**Comment by Cadair** [comment]
@pllim I have just read your GSOC idea on splitting astropy test helper out into a seperate package. This is the first step in that I think.

**Comment by pllim** [comment]
I like the concept of making things more modular. I think there was a similar request from @josePhoenix a while back. However, CIs are failing.

**Comment by Cadair** [comment]
yeah this needs some more work (it's tied to sunpy/sunpy#1983) but it's not very high on my todo list unfortunately.

**Comment by pllim** [comment]
Okay, I'll make a note of this PR number in the GSoC idea, just in case. Thanks!

**Comment by Cadair** [comment]
@pllim We moved the ideas to the open astronomy website, I added the PR and put myself up as a mentor while I was moving it ;)

**Comment by mohanagr** [comment]
@Cadair @pllim This is specifically the idea I was interested in. Can you please elaborate on how I can get involved? (As of now I'm working on [ #5799 #5858 ] ). Thanks.

**Comment by Cadair** [comment]
Well this currently breaks all the CI builds as the split clearly hasn't been done cleanly enough.

The basic idea here is to move the "plugin" code out of the `conftest.py` file and into files that can be imported separately by packages that depend on astropy (or copied out of astropy altogether) and then configure astropy to use those plugins by specifying them in the `conftest.py` file.

**Comment by mohanagr** [comment]
Also @pllim can you please tell me any small to-do's that I can begin with to start sorting this one out?

File: astropy/tests/pytest_doctestplus.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+import re
+import sys
+
+import pytest
+
+from .output_checker import AstropyOutputChecker, FIX
+
+try:
+    import importlib.machinery as importlib_machinery
+except ImportError:  # Python 2.7
```
Comment (pllim) [inline astropy/tests/pytest_doctestplus.py]: @mohanagr , Travis CI indicates failures for Python 2.7. I suspect this is the cause here. So this PR needs to be patched to work in Python 2. Also, fix PEP8 test failure.

Comment (mohanagr) [inline astropy/tests/pytest_doctestplus.py]: @pllim I'm looking through these modules. These are the ones which were separated from conftest.py. Can you tell me what all additions are already planned to be made?

Comment (pllim) [inline astropy/tests/pytest_doctestplus.py]: > Can you tell me what all additions are already planned to be made?

I am not sure if I understand the question. There is no planned addition if you mean additional keywords. But one of the proposed GSoC project is to separate out these `pytest` add-ons into a separate installable package that does not depend on Astropy.

Comment (mohanagr) [inline astropy/tests/pytest_doctestplus.py]: @pllim yes about that, doesn't this PR do that? splitting into modules? I was working on python2 patch thing.

Comment (pllim) [inline astropy/tests/pytest_doctestplus.py]: This PR simply separates them into different modules. But the modules are still part of Astropy. The GSoC project is to create a separate *package* (something parallels to Astropy, not underneath it). Does this make sense?

Comment (mohanagr) [inline astropy/tests/pytest_doctestplus.py]: Oh. Yes. Thanks. So as of now I model the changes in this PR that work with Python2 and submit another one?

Comment (pllim) [inline astropy/tests/pytest_doctestplus.py]: @mohanagr , yes, if you want to take over this work from @Cadair , you need to open a new PR from your own branch in your own fork. However, if possible, please credit @Cadair for the initial work before you apply your own changes. One way is to follow guidance given at https://github.com/astropy/astropy/pull/3558#issuecomment-287369605 (and a few comments that followed) on how to cherry pick commits from this PR. Hope this helps.

**Comment by mohanagr** [comment]
@Cadair I was a little confused as to why this project was given _Intermediate/Advanced_ on OpenAstro project list tag if plugin's code was already implemented? (I do understand that some of `astropy.utils` and `astropy.test.helpers` needs to be re-implemented for separating out the modules.)

**Comment by pllim** [comment]
@mohanagr , the hard part is disentangling dependency and making it a viable independent package. If you find it easy, then great. Or if you find it too boring, then you are free to pick a different GSoC project to apply for.

**Comment by mohanagr** [comment]
@pllim As a matter of fact I'm planning to submit for the splitting project! I have been looking into how astropy is structured and the test suites. I think there will be a lot to learn about packaging projects and making them work across different versions.

**Comment by mohanagr** [comment]
@Cadair In the new standalone, are we to provide `tests.helper` functionality like custom warnings, handling deprecations (as exceptions), handling _unicode guidelines_ etc. too? I mean features which are Astropy specific?
I mean do we single out functions/classes from `astropy.tests.helpers` and `astropy.utils` that the plugin code explicitly depends on or implement the whole thing?

**Comment by pllim** [comment]
@mohanagr , that is exactly the problem that I hope the GSoC project can solve. :sweat_smile:

**Comment by mohanagr** [comment]
There is also this problem of running all plugin code from `conftest.py` itself. This implies I can't use `--doctest-rst` option if I run from let's say `astropy` but only from `astropy/astropy` and child directories. 

This behavior was an issue with pytest.
Reference :
https://github.com/pytest-dev/pytest/issues/906
In response a warning was added regarding the use of `pytest.addoption` -
https://github.com/pytest-dev/pytest/pull/974/commits/ccf7584fac39fa2bec58e7d5d9c114dc9eb060f2

**Comment by Cadair** [comment]
@mohanagr you mean registering the plugins in `conftest` rather than installing them and having them register as hooks?

**Comment by mohanagr** [comment]
@Cadair I meant that hook calls i.e. call to `pytest_addoption` from a plugin via `pytest_plugins = ['plugin1']` might resolve the issue instead of all hook calls being in `conftest.py` itself.

This is what the comment by the maintainer reads:
> If you change your conftest file into a plugin, it should work consistently no matter where you call py.test from, or what you pass on the command-line or perhaps via `entry_points` in `setup.py`.

> It is unfortunate (but understandable) that there's this gotcha when working with pytest_addoption and conftest files, because conftest initialization being lazy as it is now can lead to surprising errors. Perhaps a note to pytest_addoption discouraging its use from conftest files would at least warn users of these pitfall?

Also, can you tell me what did you mean by _installing_ plugins? I mean if they are there in the code itself.

**Comment by mohanagr** [comment]
I was wrong above. The current way of specifying plugins in a file via `pytest_plugins=` produces the same thing (Tried changes in this PR). Declaring `entry_point` for plugins might be an option. @Cadair If you can tell me what the maintainer meant by _changing conftest file to a plugin_ I might make some headway.

**Comment by bsipocz** [comment]
> Also, can you tell me what did you mean by installing plugins? I mean if they are there in the code itself.

@mohanagr - you may want to check out other pytest plugins for the infrastructure and how-to (e.g. pytest-mpl, pytest-cov, etc). As it was said above one of the idea for the gsoc project was to factor out all the additions we have (compared to pure pytest) into an independent plugin. That means that it needs to be installed, and we don't have it any more in astropy core (pending a deprecation period of course).

**Comment by mohanagr** [comment]
@bsipocz Oh all right. Until now I was assuming there'd be separate plugins. Hence didn't understand why would one do `pip install pytest-plugin` multiple times instead of just cloning and using them.

**Comment by bsipocz** [comment]
@mohanagr - Users don't clone anything, but install released versions.

**Comment by mohanagr** [comment]
What if users are developers? :smiley:

**Comment by bsipocz** [comment]
You don't need to worry about that subset. Anyway cloning and using is not the preferred way for developers either when the aim is to build a stable package rather than hacking.

**Comment by mohanagr** [comment]
@bsipocz What then is reason behind `astropy-helpers` being used as such inside the astropy code? What if someone else needed to use them? Wouldn't he have to clone?

**Comment by pllim** [comment]
I think we are getting a bit too far off topic here. Questions related to the GSoC project should be discussed outside of this PR thread. As far as this PR is concerned, if you would like to take over, it is just simply splitting the current stuff into different modules, not a separate installable package.

**Comment by mohanagr** [comment]
@pllim I understand. I couldn't find a way to reach you apart from GH comments!

**Comment by MSeifert04** [comment]
There is a developer mailing list https://groups.google.com/forum/#!forum/astropy-dev

**Comment by bsipocz** [comment]
@Cadair - pinging you again, whether we need to remilestone this or you can make some progress with it to be merged this week?

**Comment by Cadair** [comment]
I will not be able to get to this before the freeze :(

**Comment by bsipocz** [comment]
Superseded by #6384.

**Pull request closed without merging**